File: projects/basis/basis-sdk-python/basis/modules/factory.py
import json
import os
from pathlib import Path
from web3 import Web3

def load_abi(filename: str):
    abi_path = Path(__file__).parent.parent / 'abis' / filename
    with open(abi_path, 'r', encoding='utf-8') as f:
        data = json.load(f)
        return data.get('abi', [])

class FactoryModule:
    def __init__(self, client, factory_address: str):
        self.client = client
        self.factory_address = Web3.to_checksum_address(factory_address)
        self.factory_abi = load_abi('ATokenFactory.json')
        self.token_abi = load_abi('FACTORYTOKEN.json')
        self.contract = self.client.web3.eth.contract(address=self.factory_address, abi=self.factory_abi)

    def _build_and_send_tx(self, function_call, value=0):
        if not self.client.account:
            raise ValueError("Stateful initialization (private_key) is required for write methods.")
        
        tx = function_call.build_transaction({
            'from': self.client.account.address,
            'nonce': self.client.web3.eth.get_transaction_count(self.client.account.address),
            'value': value,
            # Let Web3.py estimate gas and gas price automatically
        })

        signed_tx = self.client.web3.eth.account.sign_transaction(tx, private_key=self.client.account.key)
        tx_hash = self.client.web3.eth.send_raw_transaction(signed_tx.raw_transaction)
        receipt = self.client.web3.eth.wait_for_transaction_receipt(tx_hash)
        
        return {
            'hash': tx_hash.hex(),
            'receipt': receipt
        }

    def _create_token(
        self,
        symbol: str,
        name: str,
        hybrid_multiplier: int,
        frozen: bool,
        usdb_for_bonding: int,
        start_lp: int,
        auto_vest: bool,
        auto_vest_duration: int,
        gradual_autovest: bool
    ):
        fee_amount = self.contract.functions.feeAmount().call()
        func = self.contract.functions.createToken(
            symbol,
            name,
            hybrid_multiplier,
            frozen,
            usdb_for_bonding,
            start_lp,
            auto_vest,
            auto_vest_duration,
            gradual_autovest
        )
        return self._build_and_send_tx(func, value=fee_amount)

    def create_token_with_metadata(
        self,
        symbol: str,
        name: str,
        hybrid_multiplier: int,
        start_lp: int,
        description: str = None,
        image_url: str = None,
        website: str = None,
        telegram: str = None,
        twitterx: str = None,
        frozen: bool = False,
        usdb_for_bonding: int = 0,
        auto_vest: bool = False,
        auto_vest_duration: int = 0,
        gradual_autovest: bool = False,
    ):
        """Creates a token and registers its metadata on IPFS in one call.

        Requires SIWE authentication (call client.authenticate() first).

        1. Creates the token on-chain
        2. Parses the new token address from logs
        3. Downloads, resizes (512x512 WebP), and uploads the image to IPFS
        4. Creates metadata on IPFS

        Returns dict with hash, receipt, token_address, image_url, metadata.
        """
        # 1. Create token on-chain
        create_result = self._create_token(
            symbol=symbol, name=name,
            hybrid_multiplier=hybrid_multiplier, frozen=frozen,
            usdb_for_bonding=usdb_for_bonding, start_lp=start_lp,
            auto_vest=auto_vest, auto_vest_duration=auto_vest_duration,
            gradual_autovest=gradual_autovest,
        )

        receipt = create_result['receipt']
        if receipt.get('status') == 0:
            raise RuntimeError(f"Token creation reverted (tx: {create_result['hash']})")

        # 2. Parse token address from TokenCreated event
        token_created_topic = Web3.keccak(text="TokenCreated(address,string,string,address)").hex()
        factory_lower = self.factory_address.lower()
        token_address = None
        for log_entry in receipt.get('logs', []):
            addr = log_entry.get('address', '')
            if addr.lower() != factory_lower:
                continue
            topics = log_entry.get('topics', [])
            if not topics:
                continue
            t0 = topics[0].hex() if isinstance(topics[0], bytes) else str(topics[0])
            if t0 == token_created_topic and len(topics) > 1:
                raw = topics[1].hex() if isinstance(topics[1], bytes) else str(topics[1])
                token_address = Web3.to_checksum_address("0x" + raw[-40:])
                break

        if not token_address:
            raise RuntimeError("Could not extract token address from creation logs.")

        # 3. Upload image if provided
        uploaded_image_url = None
        if image_url:
            uploaded_image_url = self.client.api.upload_image_from_url(image_url, contract_address=token_address)

        # 4. Create metadata on IPFS
        metadata = self.client.api.update_metadata(
            address=token_address,
            description=description,
            image=uploaded_image_url,
            website=website,
            telegram=telegram,
            twitterx=twitterx,
        )

        return {
            'hash': create_result['hash'],
            'receipt': receipt,
            'token_address': token_address,
            'image_url': uploaded_image_url,
            'metadata': metadata,
        }

    def disable_freeze(self, token_address: str):
        checksum_addr = Web3.to_checksum_address(token_address)
        token_contract = self.client.web3.eth.contract(address=checksum_addr, abi=self.token_abi)
        func = token_contract.functions.DisableFreeze()
        return self._build_and_send_tx(func)

    def set_whitelisted_wallet(self, token_address: str, wallets: list[str], amount: int, tag: str):
        checksum_addr = Web3.to_checksum_address(token_address)
        token_contract = self.client.web3.eth.contract(address=checksum_addr, abi=self.token_abi)
        checksum_wallets = [Web3.to_checksum_address(w) for w in wallets]
        func = token_contract.functions.SetWhitelistedWallet(checksum_wallets, amount, tag)
        return self._build_and_send_tx(func)

    def is_ecosystem_token(self, token_address: str) -> bool:
        """Checks if a token is an ecosystem token."""
        return self.contract.functions.isEcosystemToken(Web3.to_checksum_address(token_address)).call()

    def get_tokens_by_creator(self, creator: str) -> list:
        """Returns all tokens created by a given address."""
        return self.contract.functions.getTokensByCreator(Web3.to_checksum_address(creator)).call()

    def get_fee_amount(self) -> int:
        """Returns the current fee amount for token creation."""
        return self.contract.functions.feeAmount().call()

    def remove_whitelist(self, token_address: str, wallet: str):
        """Removes a wallet from a token's whitelist."""
        checksum_token = Web3.to_checksum_address(token_address)
        checksum_wallet = Web3.to_checksum_address(wallet)
        token_contract = self.client.web3.eth.contract(address=checksum_token, abi=self.token_abi)
        func = token_contract.functions.RemoveWhitelist(checksum_wallet)
        return self._build_and_send_tx(func)

    def claim_rewards(self, token_address: str):
        """Claim accumulated USDB rewards from presale shares on a factory token."""
        checksum_addr = Web3.to_checksum_address(token_address)
        token_contract = self.client.web3.eth.contract(address=checksum_addr, abi=self.token_abi)
        func = token_contract.functions.claimRewards()
        return self._build_and_send_tx(func)

    def get_claimable_rewards(self, token_address: str, investor: str) -> int:
        """Get claimable USDB rewards for an address on a factory token."""
        checksum_addr = Web3.to_checksum_address(token_address)
        checksum_investor = Web3.to_checksum_address(investor)
        token_contract = self.client.web3.eth.contract(address=checksum_addr, abi=self.token_abi)
        return token_contract.functions.getClaimableRewards(checksum_investor).call()

    def get_token_state(self, token_address: str):
        checksum_addr = Web3.to_checksum_address(token_address)
        token_contract = self.client.web3.eth.contract(address=checksum_addr, abi=self.token_abi)
        
        frozen = token_contract.functions.frozen().call()
        has_bonded = token_contract.functions.hasBonded().call()
        total_supply = token_contract.functions.totalSupply().call()
        usd_price = token_contract.functions.getUSDPrice().call()

        return {
            'frozen': frozen,
            'hasBonded': has_bonded,
            'totalSupply': str(total_supply),
            'usdPrice': str(usd_price)
        }

File: projects/basis/sdk-review/py-v2/src/modules/Factory.ts
import { BasisClient } from '../BasisClient';
import ATokenFactoryArtifact from '../abis/ATokenFactory.json';
import FactoryTokenArtifact from '../abis/FACTORYTOKEN.json';
import { parseEther, getContract, getAddress, keccak256, toBytes, Address, GetContractReturnType } from 'viem';

export class FactoryModule {
  private client: BasisClient;
  private factoryAddress: Address;

  constructor(client: BasisClient, factoryAddress: Address) {
    this.client = client;
    this.factoryAddress = factoryAddress;
  }

  private async _syncTx(txHash: string) {
    try {
      await this.client.api.syncTransaction(txHash);
    } catch (e: any) {
      console.warn('Sync warning:', e.message || e);
    }
  }

  /**
   * Internal: creates a token on-chain. Use createTokenWithMetadata() instead.
   */
  private async createToken(
    symbol: string,
    name: string,
    hybridMultiplier: bigint,
    frozen: boolean,
    usdbForBonding: bigint,
    startLP: bigint,
    autoVest: boolean,
    autoVestDuration: bigint,
    gradualAutovest: boolean
  ) {
    if (!this.client.walletClient) {
      throw new Error("Stateful initialization (walletClient) is required for write methods.");
    }

    const account = this.client.walletClient.account;
    if (!account) throw new Error("Account is required");

    // Get the create fee from the contract
    const feeAmount = await this.client.publicClient.readContract({
      address: this.factoryAddress,
      abi: ATokenFactoryArtifact.abi,
      functionName: 'feeAmount',
    }) as bigint;

    const { request } = await this.client.publicClient.simulateContract({
      account,
      address: this.factoryAddress,
      abi: ATokenFactoryArtifact.abi,
      functionName: 'createToken',
      args: [
        symbol,
        name,
        hybridMultiplier,
        frozen,
        usdbForBonding,
        startLP,
        autoVest,
        autoVestDuration,
        gradualAutovest
      ],
      value: feeAmount,
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });

    return { hash, receipt };
  }

  /**
   * Creates a token and registers its metadata on IPFS in one call.
   * Requires SIWE authentication (call client.authenticate() first).
   *
   * 1. Creates the token on-chain
   * 2. Parses the new token address from logs
   * 3. Downloads, resizes (512x512 WebP), and uploads the image to IPFS
   * 4. Creates metadata on IPFS (name, symbol, description auto-read from chain)
   *
   * Returns { hash, receipt, tokenAddress, imageUrl, metadata }
   */
  async createTokenWithMetadata(options: {
    symbol: string;
    name: string;
    hybridMultiplier: bigint;
    frozen?: boolean;
    usdbForBonding?: bigint;
    startLP: bigint;
    autoVest?: boolean;
    autoVestDuration?: bigint;
    gradualAutovest?: boolean;
    description?: string;
    imageUrl?: string;
    website?: string;
    telegram?: string;
    twitterx?: string;
  }) {
    // 1. Create token on-chain
    const createResult = await this.createToken(
      options.symbol,
      options.name,
      options.hybridMultiplier,
      options.frozen ?? false,
      options.usdbForBonding ?? 0n,
      options.startLP,
      options.autoVest ?? false,
      options.autoVestDuration ?? 0n,
      options.gradualAutovest ?? false,
    );

    if (createResult.receipt.status === 'reverted') {
      throw new Error(`Token creation reverted (tx: ${createResult.hash})`);
    }

    // 2. Parse token address from TokenCreated event
    const TOKEN_CREATED_TOPIC = keccak256(toBytes('TokenCreated(address,string,string,address)'));
    const tokenCreatedLog = createResult.receipt.logs.find(
      (l: any) => l.address.toLowerCase() === this.factoryAddress.toLowerCase() && l.topics[0] === TOKEN_CREATED_TOPIC
    );

    let tokenAddress: Address;
    if (tokenCreatedLog && tokenCreatedLog.topics[1]) {
      tokenAddress = getAddress('0x' + tokenCreatedLog.topics[1].slice(26)) as Address;
    } else {
      throw new Error('Could not extract token address from creation logs.');
    }

    // 3. Upload image if provided
    let imageUrl: string | undefined;
    if (options.imageUrl) {
      imageUrl = await this.client.api.uploadImageFromUrl(options.imageUrl, tokenAddress);
    }

    // 4. Create metadata on IPFS
    const metadata = await this.client.api.updateMetadata({
      address: tokenAddress,
      description: options.description,
      image: imageUrl,
      website: options.website,
      telegram: options.telegram,
      twitterx: options.twitterx,
    });

    // Sync the creation tx
    this._syncTx(createResult.hash);

    return {
      hash: createResult.hash,
      receipt: createResult.receipt,
      tokenAddress,
      imageUrl,
      metadata,
    };
  }

  async disableFreeze(tokenAddress: Address) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Wallet account is required.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: tokenAddress,
      abi: FactoryTokenArtifact.abi,
      functionName: 'DisableFreeze',
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });
    this._syncTx(hash);
    return receipt;
  }

  async setWhitelistedWallet(tokenAddress: Address, wallets: Address[], amount: bigint, tag: string) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Wallet account is required.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: tokenAddress,
      abi: FactoryTokenArtifact.abi,
      functionName: 'SetWhitelistedWallet',
      args: [wallets, amount, tag],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });
    this._syncTx(hash);
    return receipt;
  }

  async getTokenState(tokenAddress: Address) {
    // We can bundle requests via multicall for better performance, but here we do simple reads for brevity
    const contract = getContract({
      address: tokenAddress,
      abi: FactoryTokenArtifact.abi,
      client: this.client.publicClient,
    });

    const [frozen, hasBonded, totalSupply, usdPrice] = await Promise.all([
      contract.read.frozen(),
      contract.read.hasBonded(),
      contract.read.totalSupply(),
      contract.read.getUSDPrice()
    ]);

    return {
      frozen: frozen as boolean,
      hasBonded: hasBonded as boolean,
      totalSupply: (totalSupply as bigint).toString(),
      usdPrice: (usdPrice as bigint).toString()
    };
  }

  /**
   * Checks if a token address belongs to the ecosystem.
   */
  async isEcosystemToken(tokenAddress: Address): Promise<boolean> {
    return this.client.publicClient.readContract({
      address: this.factoryAddress,
      abi: ATokenFactoryArtifact.abi,
      functionName: 'isEcosystemToken',
      args: [tokenAddress],
    }) as Promise<boolean>;
  }

  /**
   * Returns all token addresses created by a given creator.
   */
  async getTokensByCreator(creator: Address): Promise<Address[]> {
    return this.client.publicClient.readContract({
      address: this.factoryAddress,
      abi: ATokenFactoryArtifact.abi,
      functionName: 'getTokensByCreator',
      args: [creator],
    }) as Promise<Address[]>;
  }

  /**
   * Returns the current fee amount (in wei) required to create a token.
   */
  async getFeeAmount(): Promise<bigint> {
    return this.client.publicClient.readContract({
      address: this.factoryAddress,
      abi: ATokenFactoryArtifact.abi,
      functionName: 'feeAmount',
    }) as Promise<bigint>;
  }

  /**
   * Removes a wallet from the whitelist on a FACTORYTOKEN contract.
   */
  /**
   * Claim accumulated USDB rewards from presale shares on a factory token.
   */
  async claimRewards(tokenAddress: Address) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Wallet account is required.");
    }
    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: tokenAddress,
      abi: FactoryTokenArtifact.abi,
      functionName: 'claimRewards',
    });
    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });
    this._syncTx(hash);
    return { hash, receipt };
  }

  /**
   * Get claimable USDB rewards for an address on a factory token.
   */
  async getClaimableRewards(tokenAddress: Address, investor: Address): Promise<bigint> {
    return this.client.publicClient.readContract({
      address: tokenAddress,
      abi: FactoryTokenArtifact.abi,
      functionName: 'getClaimableRewards',
      args: [investor],
    }) as Promise<bigint>;
  }

  async removeWhitelist(tokenAddress: Address, wallet: Address) {
    if (!this.client.walletClient || !this.client.walletClient.account) {
      throw new Error("Wallet account is required.");
    }

    const { request } = await this.client.publicClient.simulateContract({
      account: this.client.walletClient.account,
      address: tokenAddress,
      abi: FactoryTokenArtifact.abi,
      functionName: 'RemoveWhitelist',
      args: [wallet],
    });

    const hash = await this.client.walletClient.writeContract(request);
    const receipt = await this.client.publicClient.waitForTransactionReceipt({ hash });
    this._syncTx(hash);
    return receipt;
  }
}
